### PR TITLE
Added transaction data builder

### PR DESCRIPTION
**Pull request by iulianpascalau**
- added transaction data builder

### CODE REVIEW

File: builders/errors.go [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package elrond
```
Comment (bogdan-rosianu) [inline builders/errors.go]: why the package is named `elrond`?

File: builders/txDataBuilder.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	}
+
+	return builder.addBytes(address.AddressBytes())
+}
+
+func (builder *txDataBuilder) checkAddress(address core.AddressHandler) error {
+	if check.IfNil(address) {
+		return fmt.Errorf("%w in builder.checkAddress", ErrNilAddress)
+	}
+	if len(address.AddressBytes()) == 0 {
```
Comment (bogdan-rosianu) [inline builders/txDataBuilder.go:101]: might be checked against the actual number of bytes for a valid address

File: builders/txDataBuilder.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+	return builder
+}
+
+func (builder *txDataBuilder) addBytes(bytes []byte) TxDataBuilder {
+	if len(bytes) == 0 {
+		bytes = []byte{0}
+	}
+
+	builder.args = append(builder.args, hex.EncodeToString(bytes))
```
Comment (bogdan-rosianu) [inline builders/txDataBuilder.go:141]: please check if the hex arguments have an even number of characters. valid for the entire struct

File: builders/txDataBuilder.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+	dataField, err := builder.ToDataString()
+	if err != nil {
+		return nil, err
+	}
+
+	return []byte(dataField), err
+}
+
+// ToVmValueRequest returns the VmValueRequest structure to be used in a VM call
+func (builder *txDataBuilder) ToVmValueRequest() (*data.VmValueRequest, error) {
```
Comment (bogdan-rosianu) [inline builders/txDataBuilder.go:168]: this is confusing. Why don't you create 2 different components? one for tx payload construction and one for vm values request builder.. or at least have 2 interfaces that are implemented by the same component (you'll have to return  `*txDataBuilder` instead of an interface. This way you'll have `TxPayloadBuilder` and `VmValuesQueryBuilder` which only expose functions correlated with their names.

If I only use your struct for preparing sc calls, seeing `ToVmValueRequest()` or `CallerAddress` might not be the base case.

File: builders/txDataBuilder.go [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+	callerAddr string
+	args       []string
+	log        logger.Logger
+	err        error
+}
+
+// NewTxDataBuilder creates a new transaction data builder
+func NewTxDataBuilder(log logger.Logger) *txDataBuilder {
+	builder := &txDataBuilder{
+		log: log,
```
Comment (bogdan-rosianu) [inline builders/txDataBuilder.go:31]: please remove the log param. this component should be simply enough so it shouldn't have any parameter on the constructor

File: builders/txDataBuilder.go [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+	builder.args = append(builder.args, hex.EncodeToString(bytes))
+
+	return builder
+}
+
+// ToDataString returns the formatted data string ready to be used in a transaction call
+func (builder *txDataBuilder) ToDataString() (string, error) {
+	if builder.err != nil {
+		return "", builder.err
```
Comment (bogdan-rosianu) [inline builders/txDataBuilder.go:149]: interesting approach. 
Suggestion: error check should be on all exported functions and only propagate the error further if any. This way, as a caller, you'll receive the earliest error so you'll keep trying from the begging to the end of the process.

Example:
```
NewDataBuilder()
   .ArgHexString("6f6b") // all good here
   .ArgAddress(nil)  // will set the error here
   .ArgHexString("6b6f") // will ignore the call with an early exit (if builder.err != nil)
   .ArgBytes([]byte{0,1,2,3}) // will ignore the call with an early exit
   .ToDataString() // will return the error from the step when setting address
```

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: builders/errors.go (hunk 1 above)
Comment (iulianpascalau) [reply builders/errors.go]: it was copied from the eth bridge. pushed before finishing the review

File: builders/txDataBuilder.go (hunk 2 above)
Comment (iulianpascalau) [reply builders/txDataBuilder.go:101]: fixed. AddressHandler already have an `IsValid` function. Used that. Thanks

File: builders/txDataBuilder.go (hunk 3 above)
Comment (iulianpascalau) [reply builders/txDataBuilder.go:141]: not required. Please test this:
```GO
func TestHex(t *testing.T){
    b := []byte{1}
    assert.Equal(t, 2, len(hex.EncodeToString(b)))
}
```

File: builders/txDataBuilder.go (hunk 4 above)
Comment (iulianpascalau) [reply builders/txDataBuilder.go:168]: split in 2. Extracted common code as maximum it can get.

File: builders/txDataBuilder.go (hunk 5 above)
Comment (iulianpascalau) [reply builders/txDataBuilder.go:31]: removed, unused

File: builders/txDataBuilder.go (hunk 6 above)
Comment (iulianpascalau) [reply builders/txDataBuilder.go:149]: done in the baseBuilder struct

**Review by bogdan-rosianu — APPROVED** [review]
:+1: only thing needed: unit tests that ensure that the first encountered error is the one returned in the end

File: storage/mapCacher.go [hunk 7]
```diff
@@ -8,7 +8,7 @@ import (
 	logger "github.com/ElrondNetwork/elrond-go-logger"
 )
 
-var log = logger.GetOrCreate("mapCacher")
+var log = logger.GetOrCreate("elrond-sdk-erdgo/storage")
```
Comment (dragos-rebegea) [inline storage/mapCacher.go:11]: why did you change mapCacher to storage if the struct name has been kept the same ?

**Review by dragos-rebegea — APPROVED** [review]

File: storage/mapCacher.go (hunk 7 above)
Comment (iulianpascalau) [reply storage/mapCacher.go:11]: this is actually the package: `elrond-sdk-erdgo/storage`. The naming is now consistent in this repo. I've chosen to add the `elrond-sdk-erdgo` in order to better localize the possible problem if we used the sdk in other problem

**Review by bogdan-rosianu — APPROVED** [review]
